### PR TITLE
Add Support for Serializing Operators

### PR DESCRIPTION
**Pull request by brownj85**
Note to reviewers: Feedback on docstrings and tests is welcome, but the main purpose of this PR is to introduce and get feedback on the serialization API. Documentation and test coverage will be brought up to the Pennylane standard before merging the feature branch in https://github.com/PennyLaneAI/pennylane/pull/4097.

**Context:**
Pennylane datasets need to support ``Operator`` classes as attributes.

**Description of the Change:**
Adds a ``DatasetOperator`` attribute type for operators. This is done by serializing the `data` attribute as an HDF5 array, and the ``wires`` attribute as JSON, and passing these to the operator classes' `__init__` method when de-serializing.

 As a result, this class only supports operators that maintain the signature of ``Operator.__init__``, and do not mutate the ``params`` argument before passing it into ``super()``.  Almost all operators meet this requirement, and ``Tensor`` and ``Hamiltonian`` are also supported as a special case.

This does mean that operator wires must be JSON serializable. This appears to be the most common case, but feedback on this issue is needed.

**Benefits:**
The majority of operators, including nested compound operators, can be serialized without pickle.

**Possible Drawbacks:**
Wires must be JSON

**Related GitHub Issues:**

### CODE REVIEW

**Comment by timmysilv** [comment]
#3970 (and the linked ADR) might be of relevance here. There is a plan in the PL roadmap to make all Operators have a PyTree representation, which would come with serialization. Regardless of what happens there, I'd like to state that I don't want multiple ways of serialization in PennyLane (not from scratch, at least). That push will involve making Operators immutable, which will be nice:)

File: pennylane/data/attributes/operator/operator.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+                qml.U2,
+                qml.U3,
+                # pennylane/ops/qubit/qchem_qml.py
+                qml.SingleExcitation,
+                qml.SingleExcitationMinus,
+                qml.SingleExcitationPlus,
+                qml.DoubleExcitation,
+                qml.DoubleExcitationMinus,
+                qml.DoubleExcitationPlus,
+                qml.OrbitalRotation,
```
Comment (obliviateandsurrender) [inline pennylane/data/attributes/operator/operator.py:123]: `qml.FermionicSWAP` seems missing, is it not serializable?

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Leaving some quick comments for now. I will probably go through the rest tomorrow.

File: pennylane/data/attributes/operator/operator.py
Comment (timmysilv) [inline pennylane/data/attributes/operator/operator.py:1]: What are the operators that can't be serialized? I'm assuming it's everything in `pennylane/ops/op_math`, but those are all heavily used. `Hamiltonian` and `Tensor` are being replaced with arithmetic operators (`Sum`, `Prod`, `SProd`), and other operators like `Pow`, `Adjoint`, `Exp`, and all the new pulse stuff will need support as well

**Review by timmysilv — COMMENTED** [review]
I'm wondering how much of this can be put beside the operators themselves rather than tucked away in this `data` module. Again, there is a serialization format coming, and I know it won't be timely enough, but having two different serialization things will be a lot of overhead maintenance for whoever is responsible for that (which I imagine will be the PL Core squad). I think these things could complement each other, so it's more of a comment than a concern, but I do think it should be considered. Especially because we need to support the more flexible arithmetic ops as well as Hamiltonian/Tensor.

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]

File: tests/data/attributes/operator/test__wires.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
+import pytest
+
+from pennylane.data.attributes.operator._wires import wires_to_json
+from pennylane.wires import Wires
+import json
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply tests/data/attributes/operator/test__wires.py:9]: I think the `np.array([0, 1, 2])` covers that 😄 
```pycon
>>> isinstance(np.array([0,1,2])[0], numbers.Integral)
True
```
though I learned `np.array(0)` is not (in other words, np.int64 is, np.ndarray is not).

File: pennylane/data/attributes/operator/operator.py
Comment (brownj85) [reply pennylane/data/attributes/operator/operator.py:1]: @obliviateandsurrender

The operators that mutate `params` or `wires` before passing them to `Operator.__init__()` aren't supported. So under the `pennylane.ops` module this is:

- `ops.qutrit.non_parametric_ops.THadamard`
- `ops.qubit.parametric_ops_multi_qubit.PauliRot`
- `ops.qubit.special_unitary.TmpPauliRot`
- `ops.qubit.matrix_ops.BlockEncode`
- `ops.qubit.non_parametric_ops.ISWAP`
- `ops.channel.QubitChannel`
- `ops.qutrit.parametric_ops.TRX`
- `ops.qubit.observables.SparseHamiltonian`
- `ops.qubit.non_parametric_ops.Barrier`
- `ops.qubit.non_parametric_ops.MultiControlledX`
- `ops.qutrit.matrix_ops.ControlledQutritUnitary`
- `ops.snapshot.Snapshot`
- `ops.qutrit.observables.GellMann`
- `ops.qubit.qchem_ops.FermionicSWAP`
- `ops.qubit.parametric_ops_multi_qubit.PCPhase`
- `ops.qubit.arithmetic_ops.IntegerComparator`
- `ops.cv.CubicPhase`

Support for `op_math` is definitely doable, but I wanted to hold of on that because priority right now is supporting existing datasets, which use `Hamiltonian` and `Tensor`. What's the pulse stuff?

File: pennylane/data/attributes/operator/operator.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        - The ``__init__()`` method matches the signature of ``Operator.__init__``,
+            or any additional arguments are optional and do not affect the value of
+            the operator
+        - The ``data`` and ``wires`` attributes will produce an identical copy of
+            operator if passed into the classes' ``__init__()`` method. Generally,
+            this means ``__init__()`` do not mutate the ``params`` and ``wires``
+            arguments.
+        - Hyperparameters are not used or are automatically derived by ``__init__()``.
+
+    Almost all operators meet these conditions. This type also supports serializing the
```
[the remark this answers is not in the document]
Comment (brownj85) [reply pennylane/data/attributes/operator/operator.py:48]: Commented above, but there are a lot more operators defined outside of `pennylane.ops` than I first realized, so I don't think it's feasible to keep an exhaustive list here. I

File: pennylane/data/attributes/operator/operator.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    ``Hamiltonian`` and ``Tensor`` operators.
+    """
+
+    type_id = "operator"
+
+    @classmethod
+    @lru_cache(1)
+    def consumes_types(cls) -> FrozenSet[Type[Operator]]:
+        return frozenset(
```
[the remark this answers is not in the document]
Comment (brownj85) [reply pennylane/data/attributes/operator/operator.py:57]: I agree, a manual list wasn't my first choice, but I think it's necessary for the MVP. From what I can tell there's no universal representation available on all operators. The combination of `data, wires` works for most of them, but the only way to check that is manually verifying.

If the Pytrees work goes ahead this won't be a problem.

File: pennylane/data/attributes/operator/operator.py
Comment (timmysilv) [reply pennylane/data/attributes/operator/operator.py:1]: Pulse is stuff like `ParametrizedHamiltonian`, see [these docs](https://docs.pennylane.ai/en/stable/code/qml_pulse.html) for more details.

That's fair re: op_math stuff, but it's also worth noting that while things like `qml.PauliZ(0) @ qml.PauliX(1)` return a `Tensor` by default (toggle available) today, the behaviour will change to return a `Prod` instead in the semi-near future (probably in 2 releases if I had to guess)

File: pennylane/data/attributes/operator/operator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/data/attributes/operator/operator.py:48]: if you want a whole list:
```python
import pennylane as qml

def recurse(c):
    subs = c.__subclasses__()
    name = c.__name__
    if not subs:
        return name
    return {name: [recurse(i) for i in subs]}

recurse(qml.operation.Operator)
```

Most non-leaf values are abstract (or at least, they are in spirit), except for a handful (usually bc of Controlled, qutrit or pulse stuff)

**Comment by brownj85** [comment]
> I'm wondering how much of this can be put beside the operators themselves rather than tucked away in this `data` module. Again, there is a serialization format coming, and I know it won't be timely enough, but having two different serialization things will be a lot of overhead maintenance for whoever is responsible for that (which I imagine will be the PL Core squad). I think these things could complement each other, so it's more of a comment than a concern, but I do think it should be considered. Especially because we need to support the more flexible arithmetic ops as well as Hamiltonian/Tensor.

Thanks for the review - totally agree that the `data` module shouldn't be responsible for the implementation of operator serialization. I don't think this solution should be maintained long-term, or even medium-term. That effort would be better put towards adding a serialization API to the `Operator` class.  We can definitely add support for `op_math` soon.

File: pennylane/data/attributes/operator/operator.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+    Supports all operator types that meet the following conditions:
+        - The ``__init__()`` method matches the signature of ``Operator.__init__``,
+            or any additional arguments are optional and do not affect the value of
+            the operator
+        - The ``data`` and ``wires`` attributes will produce an identical copy of
+            operator if passed into the classes' ``__init__()`` method. Generally,
+            this means ``__init__()`` do not mutate the ``params`` and ``wires``
+            arguments.
+        - Hyperparameters are not used or are automatically derived by ``__init__()``.
```
Comment (albi3ro) [inline pennylane/data/attributes/operator/operator.py:46]: Is there a particular difficulty in being able to serialize hyperparameters?  Would there be a representation for "non trainable information" that would allow it to be nicely serialized?

File: pennylane/data/attributes/operator/_wires.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    order of their index.
+
+    Returns:
+        JSON list of wires
+
+    Raises:
+        UnserializableWireError, if any of the wires are not JSON-serializable.
+    """
+    jsonable_wires = []
+    for w in wires:
```
[the remark this answers is not in the document]
Comment (brownj85) [reply pennylane/data/attributes/operator/_wires.py:46]: good point, cleaned up the logic for this

File: pennylane/data/attributes/operator/operator.py (hunk 5 above)
Comment (brownj85) [reply pennylane/data/attributes/operator/operator.py:46]: Probably not, I hadn't considered the issue too closely since it we didn't need to for the current datasets. The only potential issue whether the the types of hyperparameters are generally constrained (e.g JSON types, numbers, 1D arrays), or they can be any Python object. The latter case will be more difficult

File: pennylane/data/attributes/operator/operator.py (hunk 1 above)
Comment (brownj85) [reply pennylane/data/attributes/operator/operator.py:123]: It's serializable, added

File: pennylane/data/attributes/operator/operator.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                ops.append(
+                    qml.Hamiltonian(
+                        coeffs=list(bind[op_key]["hamiltonian_coeffs"]),
+                        observables=self._hdf5_to_ops(bind[op_key]),
+                    )
+                )
+            else:
+                wire_labels = json.loads(op_wire_labels[i])
+                op_data = bind[op_key]
+                if op_data.shape is not None:
```
Comment (timmysilv) [inline pennylane/data/attributes/operator/operator.py:252]: have you not hit any issues here? depending on the interface used (eg. none), `op.data` might not have a `shape` attribute:
```pycon
>>> qml.RX(1.1, 0).data.shape
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'shape'
```

`qml.math.shape` should work if it's needed though:
```pycon
>>> qml.math.shape(qml.RX(1.1, 0).data)
(1,)
```

tl;dr could you add tests that initialize a parametrized Operator like RX (eg. something other than Hamiltonian) using vanilla python values like 1.1 above, as well as using a numpy value that has a shape attribute?

**Review by timmysilv — COMMENTED** [review]
looks super good - just one more concern, but after that this is probably ready to go!

ps - creating `ham_grp`, setting a value then never using it looked really strange but I realize it's changing memory behind the scenes that can be accessed later 😅

File: pennylane/data/attributes/operator/operator.py (hunk 7 above)
Comment (brownj85) [reply pennylane/data/attributes/operator/operator.py:252]: In this case, `op_data` is an HDF5 array, it always has a `shape` attribute. HDF5 arrays only have `shape=None` if they are explicitly defined as 'Empty'. This is distinct from a zero-length array:
```
>>> f = h5py.File("tmp", "w")
>>> f["empty"] = h5py.Empty("f")
>>> f["empty"].shape
None
>>> f["zero_length"] = []
>>> f["zero_length"].shape
(0,)
```

I'll add some tests for the RX operator
